docs/architecture/adr-049-state-sync-hooks.md

```diff
@@ -0,0 +1,187 @@
+the IAVL tree, they can not add their states to the snapshot stream for state-sync.
+
+Notification: `Even though the module state data is outside of the tree, for determinism we
+require that the hash of the external data should be posted in the IAVL tree`
+
+## Decision
+
+A simple proposal based on our existing implementation is that, we can add two new message types: `SnapshotExtensionMeta` 
+and `SnapshotExtensionPayload`, and they are appended to the existing multi-store stream with `SnapshotExtensionMeta` 
+acting as a delimiter between extensions. As the chunk hashes should be able to ensure data integrity, we don't need 
+a delimiter to mark the end of the snapshot stream.
+
+Besides, we provide `Snapshotter` and `ExtensionSnapshotter` interface for modules to implement snapshotters, which will handle both taking 
+snapshot and the restoration. Each module could have mutiple snapshotters, and for modules with additional state, they should
+implement `ExtensionSnapshotter` as extension snapshotters. When setting up the application, the snapshot `Manager` should call 
+`RegisterExtensions([]ExtensionSnapshotter…)` to register all the extension snapshotters.
+
+```proto
+// SnapshotItem is an item contained in a rootmulti.Store snapshot.
+message SnapshotItem {
+  // item is the specific type of snapshot item.
+  oneof item {
+    SnapshotStoreItem        store             = 1;
+    SnapshotIAVLItem         iavl              = 2 [(gogoproto.customname) = "IAVL"];
+    SnapshotExtensionMeta    extension         = 3;
+    SnapshotExtensionPayload extension_payload = 4;
+  }
+}
+
+// SnapshotStoreItem contains metadata about a snapshotted store.
+message SnapshotStoreItem {
+  string name = 1;
+}
+
+// SnapshotIAVLItem is an exported IAVL node.
+message SnapshotIAVLItem {
+  bytes key     = 1;
+  bytes value   = 2;
+  int64 version = 3;
+  int32 height  = 4;
+}
+
+// SnapshotExtensionMeta contains metadata about an external snapshotter.
+// One module may need multiple snapshotters, so each module may have multiple SnapshotExtensionMeta.
+message SnapshotExtensionMeta {
+  string name   = 1;
+  // format is used within the snapshotter/namespace, not global one for all modules
+  uint32 format = 2;
+}
+
+// SnapshotExtensionPayload contains payloads of an external snapshotter.
+message SnapshotExtensionPayload {
+  bytes payload = 1;
+}
+```
+
+The snapshot stream would look like this:
+
+```go
+// multi-store snapshot
+{SnapshotStoreItem | SnapshotIAVLItem, ...}
+// extension1 snapshot
+SnapshotExtensionMeta
+{SnapshotExtensionPayload, ...}
+// extension2 snapshot
+SnapshotExtensionMeta
+{SnapshotExtensionPayload, ...}
+```
+
+Add `extensions` field to snapshot `Manager` for extension snapshotters. `multistore` snapshotter is a special one and it doesn't need a name because it is always placed at the beginning of the binary stream.
```

Comment:
```suggestion
We add an `Extensions` field to snapshot `Manager` for extension snapshotters. The `multistore` snapshotter is a special one and it doesn't need a name because it is always placed at the beginning of the binary stream.
```